### PR TITLE
[wip] building a sea wall for atlantis

### PR DESCRIPTION
**Pull request by jonpulsifer**
Atlantis AppEngine support: https://github.com/jonpulsifer/atlantis/pull/1
AppEngine Flex tf module: https://github.com/jonpulsifer/terraform-modules/pull/1
Robots fighting: https://github.com/jonpulsifer/cloudlab/pull/14

### CODE REVIEW

**Comment by j0npulsifer** [comment]
Ran Plan for dir: `gcp/projects/sea-wall` workspace: `default`

**Plan Error**
<details><summary>Show Output</summary>

```
running "/usr/local/bin/terraform init -input=false -no-color -upgrade" in "/home/atlantis/.atlantis/repos/jonpulsifer/cloudlab/19/default/gcp/projects/sea-wall": exit status 1
Upgrading modules...
- atlantis in 
- network in 

Error: Unreadable module directory

Unable to evaluate directory symlink: lstat ../../../../terraform-modules: no
such file or directory


Error: Failed to read module directory

Module directory  does not exist or cannot be read.


Error: Unreadable module directory

Unable to evaluate directory symlink: lstat ../../../../terraform-modules: no
such file or directory


Error: Failed to read module directory

Module directory  does not exist or cannot be read.


Error: Unreadable module directory

Unable to evaluate directory symlink: lstat ../../../../terraform-modules: no
such file or directory


Error: Failed to read module directory

Module directory  does not exist or cannot be read.


Error: Unreadable module directory

Unable to evaluate directory symlink: lstat ../../../../terraform-modules: no
such file or directory


Error: Failed to read module directory

Module directory  does not exist or cannot be read.


Error: Unreadable module directory

Unable to evaluate directory symlink: lstat ../../../../terraform-modules: no
such file or directory


Error: Failed to read module directory

Module directory  does not exist or cannot be read.


```
</details>

File: gcp/projects/sea-wall/gce.tf [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module "network" {
+  source = "../../../../terraform-modules/gce-vpc"
```
Comment (jonpulsifer) [inline gcp/projects/sea-wall/gce.tf]: move to gcs once the module is there

File: gcp/projects/seawall-ae-1234567/policy.tf [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  project = local.project
+  for_each = {
+    "compute.trustedImageProjects" : [
+      "projects/gce-uefi-images",
+      "projects/cos-cloud",
+    ],
+    "gcp.resourceLocations" : [
+      #"in:northamerica-northeast1-locations",
+      "in:northamerica-locations", # storage is global :( (gcr)
+    ],
```
Comment (jonpulsifer) [inline gcp/projects/seawall-ae-1234567/policy.tf]: see if we can pin gcr to canada?

File: gcp/projects/seawall-ae-1234567/versions.tf [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+provider "google-beta" {
+  project = local.project
+  region  = local.region
+  zone    = local.zone
+}
+
+terraform {
+  backend "local" {}
+}
```
Comment (jonpulsifer) [inline gcp/projects/seawall-ae-1234567/versions.tf]: move to gcs

File: gcp/projects/seawall-ae-1234567/gae.tf [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      source_range = "185.199.108.0/22"
+      priority     = 1012
+    },
+    "github-webhook-3" = {
+      action       = "ALLOW"
+      source_range = "140.82.112.0/20"
+      priority     = 1013
+    },
+    "default-tf-action" = {
+      action       = "ALLOW"
```
Comment (jonpulsifer) [inline gcp/projects/seawall-ae-1234567/gae.tf]: DENY lol https://seawall.umadbro.dev

File: gcp/projects/sea-wall/gce.tf [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  name   = "atlantis"
+}
+
+module "atlantis" {
+  source             = "../../../../terraform-modules/gce-igm-shielded"
```
Comment (jonpulsifer) [inline gcp/projects/sea-wall/gce.tf]:
```suggestion
  source             = "github.com/jonpulsifer/terraform-modules//gce-igm-shielded"
```

File: gcp/projects/seawall-ae-1234567/gae.tf [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  }
+  domain_names = {
+    "umadbro.dev"         = {},
+    "www.umadbro.dev"     = {},
+    "seawall.umadbro.dev" = {}
+  }
+}
+
+module "network" {
+  source                  = "../../../../terraform-modules/gce-vpc"
```
Comment (jonpulsifer) [inline gcp/projects/seawall-ae-1234567/gae.tf]:
```suggestion
  source                  = "github.com/jonpulsifer/terraform-modules//gce-vpc"
```

File: gcp/projects/seawall-ae-1234567/gae.tf [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  name                    = "default"
+  auto_create_subnetworks = true # gui cheat codes
+  subnet_name             = "default"
+  ip_cidr_range           = "10.162.0.0/20"
+  private_api_access      = false
+}
+
+module "atlantis" {
+  source         = "../../../../terraform-modules/appengine-flex"
```
Comment (jonpulsifer) [inline gcp/projects/seawall-ae-1234567/gae.tf]:
```suggestion
  source         = "github.com/jonpulsifer/terraform-modules//appengine-flex"
```

**Comment by j0npulsifer** [comment]
Ran Plan for dir: `gcp/projects/sea-wall` workspace: `default`

<details><summary>Show Output</summary>

```diff

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
+ create
 <= read (data resources)

Terraform will perform the following actions:

  # data.google_iam_policy.gcs-atlantis will be read during apply
  # (config refers to values not yet known)
 <= data "google_iam_policy" "gcs-atlantis"  {
      + id          = (known after apply)
      + policy_data = (known after apply)

      + binding {
          + members = [
              + "group:cloud@pulsifer.ca",
              + (known after apply),
            ]
          + role    = "roles/storage.admin"
        }
    }

  # google_project_organization_policy.bool-policies["compute.disableGuestAttributesAccess"] will be created
+ resource "google_project_organization_policy" "bool-policies" {
      + constraint  = "constraints/compute.disableGuestAttributesAccess"
      + etag        = (known after apply)
      + id          = (known after apply)
      + project     = "sea-wall"
      + update_time = (known after apply)
      + version     = (known after apply)

      + boolean_policy {
          + enforced = true
        }
    }

  # google_project_organization_policy.bool-policies["iam.disableServiceAccountCreation"] will be created
+ resource "google_project_organization_policy" "bool-policies" {
      + constraint  = "constraints/iam.disableServiceAccountCreation"
      + etag        = (known after apply)
      + id          = (known after apply)
      + project     = "sea-wall"
      + update_time = (known after apply)
      + version     = (known after apply)

      + boolean_policy {
          + enforced = false
        }
    }

  # google_project_organization_policy.bool-policies["iam.disableServiceAccountKeyCreation"] will be created
+ resource "google_project_organization_policy" "bool-policies" {
      + constraint  = "constraints/iam.disableServiceAccountKeyCreation"
      + etag        = (known after apply)
      + id          = (known after apply)
      + project     = "sea-wall"
      + update_time = (known after apply)
      + version     = (known after apply)

      + boolean_policy {
          + enforced = false
        }
    }

  # google_project_organization_policy.list-policies-allow["compute.vmExternalIpAccess"] will be created
+ resource "google_project_organization_policy" "list-policies-allow" {
      + constraint  = "constraints/compute.vmExternalIpAccess"
      + etag        = (known after apply)
      + id          = (known after apply)
      + project     = "sea-wall"
      + update_time = (known after apply)
      + version     = (known after apply)

      + list_policy {
          + inherit_from_parent = false
          + suggested_value     = (known after apply)

          + allow {
              + all = true
            }
        }
    }

  # google_project_organization_policy.list-policies-allow["serviceuser.services"] will be created
+ resource "google_project_organization_policy" "list-policies-allow" {
      + constraint  = "constraints/serviceuser.services"
      + etag        = (known after apply)
      + id          = (known after apply)
      + project     = "sea-wall"
      + update_time = (known after apply)
      + version     = (known after apply)

      + list_policy {
          + inherit_from_parent = false
          + suggested_value     = (known after apply)

          + allow {
              + all = true
            }
        }
    }

  # google_project_organization_policy.list-policies-values["compute.trustedImageProjects"] will be created
+ resource "google_project_organization_policy" "list-policies-values" {
      + constraint  = "constraints/compute.trustedImageProjects"
      + etag        = (known after apply)
      + id          = (known after apply)
      + project     = "sea-wall"
      + update_time = (known after apply)
      + version     = (known after apply)

      + list_policy {
          + inherit_from_parent = true
          + suggested_value     = (known after apply)

          + allow {
              + all    = false
              + values = [
                  + "projects/gce-uefi-images",
                ]
            }
        }
    }

  # google_project_organization_policy.list-policies-values["gcp.resourceLocations"] will be created
+ resource "google_project_organization_policy" "list-policies-values" {
      + constraint  = "constraints/gcp.resourceLocations"
      + etag        = (known after apply)
      + id          = (known after apply)
      + project     = "sea-wall"
      + update_time = (known after apply)
      + version     = (known after apply)

      + list_policy {
          + inherit_from_parent = true
          + suggested_value     = (known after apply)

          + allow {
              + all    = false
              + values = [
                  + "in:northamerica-northeast1-locations",
                ]
            }
        }
    }

  # google_project_service.compute will be created
+ resource "google_project_service" "compute" {
      + disable_dependent_services = true
      + disable_on_destroy         = true
      + id                         = (known after apply)
      + project                    = (known after apply)
      + service                    = "compute.googleapis.com"
    }

  # google_service_account.atlantis will be created
+ resource "google_service_account" "atlantis" {
      + account_id   = "atlantis"
      + display_name = "atlantis robot that runs terraform"
      + email        = (known after apply)
      + id           = (known after apply)
      + name         = (known after apply)
      + project      = (known after apply)
      + unique_id    = (known after apply)
    }

  # google_storage_bucket.atlantis will be created
+ resource "google_storage_bucket" "atlantis" {
      + bucket_policy_only = true
      + force_destroy      = false
      + id                 = (known after apply)
      + location           = "NORTHAMERICA-NORTHEAST1"
      + name               = "rusty-sea-can"
      + project            = (known after apply)
      + self_link          = (known after apply)
      + storage_class      = "STANDARD"
      + url                = (known after apply)
    }

  # google_storage_bucket_iam_policy.gcs-atlantis will be created
+ resource "google_storage_bucket_iam_policy" "gcs-atlantis" {
      + bucket      = "rusty-sea-can"
      + etag        = (known after apply)
      + id          = (known after apply)
      + policy_data = (known after apply)
    }

  # module.atlantis.google_compute_instance_group_manager.igm will be created
+ resource "google_compute_instance_group_manager" "igm" {
      + base_instance_name = "atlantis"
      + fingerprint        = (known after apply)
      + id                 = (known after apply)
      + instance_group     = (known after apply)
      + instance_template  = (known after apply)
      + name               = "atlantis-igm"
      + project            = (known after apply)
      + self_link          = (known after apply)
      + target_size        = 1
      + update_strategy    = (known after apply)
      + wait_for_instances = false
      + zone               = (known after apply)

      + update_policy {
          + max_surge_fixed       = (known after apply)
          + max_unavailable_fixed = 1
          + minimal_action        = "REPLACE"
          + type                  = "PROACTIVE"
        }

      + version {
          + instance_template = (known after apply)
          + name              = "atlantis"
        }
    }

  # module.atlantis.google_compute_instance_template.shielded_vm will be created
+ resource "google_compute_instance_template" "shielded_vm" {
      + can_ip_forward       = false
      + description          = "GCE Shielded VM Instance Template [terraform]"
      + id                   = (known after apply)
      + instance_description = "Shielded VM"
      + machine_type         = "n1-standard-1"
      + metadata             = {
          + "disable-legacy-endpoints" = "TRUE"
          + "enable-oslogin"           = "TRUE"
          + "enable-oslogin-2fa"       = "TRUE"
          + "user-data"                = <<~EOT
                #cloud-config
            EOT
        }
      + metadata_fingerprint = (known after apply)
      + name                 = (known after apply)
      + name_prefix          = "atlantis"
      + project              = (known after apply)
      + region               = (known after apply)
      + self_link            = (known after apply)
      + tags                 = [
          + "cos-beta",
          + "managed",
          + "shielded",
          + "terraform",
          + "vm",
        ]
      + tags_fingerprint     = (known after apply)

      + disk {
          + auto_delete  = true
          + boot         = true
          + device_name  = "boot"
          + disk_size_gb = 24
          + disk_type    = "pd-standard"
          + interface    = (known after apply)
          + mode         = (known after apply)
          + source_image = "https://www.googleapis.com/compute/v1/projects/gce-uefi-images/global/images/cos-beta-81-12871-44-0"
          + type         = "PERSISTENT"
        }

      + network_interface {
          + name               = (known after apply)
          + network            = (known after apply)
          + subnetwork         = (known after apply)
          + subnetwork_project = (known after apply)
        }

      + scheduling {
          + automatic_restart   = false
          + on_host_maintenance = "TERMINATE"
          + preemptible         = true
        }

      + service_account {
          + email  = (known after apply)
          + scopes = [
              + "https://www.googleapis.com/auth/cloud-platform",
            ]
        }

      + shielded_instance_config {
          + enable_integrity_monitoring = true
          + enable_secure_boot          = true
          + enable_vtpm                 = true
        }
    }

  # module.atlantis.google_service_account.igm will be created
+ resource "google_service_account" "igm" {
      + account_id   = "atlantis-igm"
      + display_name = "service account for atlantis GCE Shielded VM"
      + email        = (known after apply)
      + id           = (known after apply)
      + name         = (known after apply)
      + project      = (known after apply)
      + unique_id    = (known after apply)
    }

  # module.network.google_compute_network.network will be created
+ resource "google_compute_network" "network" {
      + auto_create_subnetworks         = false
      + delete_default_routes_on_create = false
      + gateway_ipv4                    = (known after apply)
      + id                              = (known after apply)
      + ipv4_range                      = (known after apply)
      + name                            = "atlantis"
      + project                         = (known after apply)
      + routing_mode                    = (known after apply)
      + self_link                       = (known after apply)
    }

  # module.network.google_compute_subnetwork.subnet will be created
+ resource "google_compute_subnetwork" "subnet" {
      + creation_timestamp       = (known after apply)
      + enable_flow_logs         = (known after apply)
      + fingerprint              = (known after apply)
      + gateway_address          = (known after apply)
      + id                       = (known after apply)
      + ip_cidr_range            = "10.13.37.0/28"
      + name                     = "subnet"
      + network                  = (known after apply)
      + private_ip_google_access = true
      + project                  = (known after apply)
      + purpose                  = (known after apply)
      + region                   = (known after apply)
      + secondary_ip_range       = (known after apply)
      + self_link                = (known after apply)
    }

Plan: 16 to add, 0 to change, 0 to destroy.

```

* :arrow_forward: To **apply** this plan, comment:
    * `atlantis apply -d gcp/projects/sea-wall`
* :put_litter_in_its_place: To **delete** this plan click [here](http://32bcfd5ab4e6:8080/lock?id=jonpulsifer%252Fcloudlab%252Fgcp%252Fprojects%252Fsea-wall%252Fdefault)
* :repeat: To **plan** this project again, comment:
    * `atlantis plan -d gcp/projects/sea-wall`
</details>

---
* :fast_forward: To **apply** all unapplied plans from this pull request, comment:
    * `atlantis apply`

**Pull request merged**